### ISSUE TITLE
[ENH] discoverable and searchable datasets with taxonomy

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#4332]
Following ideas of @JonathanBechtel in #4314, it might be good to make datasets discoverable by tags, e.g., is it classification or forecasting, how many classes, instances, etc.

Here's an overview table that @JonathanBechtel suggested in #4314:
![image](https://user-images.githubusercontent.com/7985502/225168189-9e983470-0947-4f90-abf5-4719c25b0a44.png)

Citing @JonathanBechtel:

> Yes, correct.  I'll probably create a new issue for this broader agenda to touch up the datasets section, once it's specced out a bit more.  There are a number of tweaks that could be made in order to make it more usable.  

>In addition to what you mentioned, here are some other items to think about:
> - Some of the datasets return data in an inconsistent format:  some datasets return `(y, X)`, and others return `(X, y)`
> - Some have a `split` argument, but I'm not sure why others wouldn't.  
> - In general the top level documentation is a little difficult to parse.  I found myself having a hard time getting around to the "right" dataset to use for an example, only to find that it didn't exist.  
> - The UCR_UEA dataset isn't actually a dataset at all, but a loader for any number of datasets from their repository, so it's mistitled in my opinion, and should be changed to `fetch_ucr_uea_dataset` or something similar to denote that it's not a pre-installed dataset but something that actively downloads new data.

> Regarding taxonomy, I think a simple improvement would be to have something on the main datasets page akin to this document:  https://docs.google.com/spreadsheets/d/1SNbdgWUoot1tSnuKlxbhKvalYti7L1Z73hnUPXAaZdg/edit?usp=sharing

> We can quibble about the details of what to include, but that seems like it'd be an improvement over what's currently here:  http://www.sktime.net/en/latest/api_reference/datasets.html, which is a little sparse on the information you'd actually need to use for an example.  I found myself having to manually download a lot of different ones to verify what was actually inside them.

My reaction to this:

If the datasets were objects, and the properties in the columns were tags, then the existing `all_estimators` could be leveraged to produce this table quite easily, by running

```python
all_estimators("dataset", as_dataframe=True, return_tags=tag_list)
```
where `tag_list` is a list of tags corresponding to the table columns.

**Comment by felipeangelimvieira** [comment]
Hey, I'm updating this issue with the current design proposal. If needed, I can also document it in the [[enhancement proposals repository](https://github.com/sktime/enhancement-proposals/)](https://github.com/sktime/enhancement-proposals/).

# Summary of Design Proposal

## Overall Dataset Interface

- One `BaseDataset` class, plus specific classes for each time series problem (e.g., `BaseForecastingDataset`, `BaseClassificationDataset`).
- All information that defines the dataset’s size and content should be provided in `__init__`.  
  The `load` method receives, as `*args`, strings that indicate which sets the method should return. For example:
  ```python
  X, y, X_train = dataset.load("X", "y", "X_train")
  ```
  If an argument does not correspond to a valid set, a `ValueError` is raised. For example, if the user tries to load `X_train` or `X_test` from a dataset that does not have train-test splits, the exception is raised.
- If the dataset has a default cross-validation split, then `X_train`, `y_train`, `X_test`, and `y_test` should be iterators containing the respective fold data. Their lengths would be equal to the number of CV folds. Each item in the `y_train`/`y_test` and `X_train`/`X_test` iterators corresponds to the training (or testing) subset for the current fold.

## Classification Tags

- **is_univariate**: bool, default=True  
  Indicates whether the dataset is univariate. In the case of a classification dataset, this refers to the dimensionality of the X dataframe (i.e., how many series are associated with each class label).
- **n_instances**: int, default=None  
  The number of instances in the dataset. Should be equal to `y.shape[0]`.
- **n_instances_train**: int, default=None  
  The number of instances in the training set. Should be equal to `y_train.shape[0]`.
- **n_instances_test**: int, default=None  
  The number of instances in the test set. Should be equal to `y_test.shape[0]`.
- **n_classes**: int, default=2  
  The number of classes in the dataset.

## Regression Tags

- **is_univariate**: bool, default=True  
  Indicates whether the dataset is univariate. In the case of a regression dataset, this refers to the dimensionality of the X dataframe (i.e., how many series are associated with each label).
- **n_instances**: int, default=None  
  The number of instances in the dataset.
- **n_instances_train**: int, default=None  
  The number of instances in the training set.
- **n_instances_test**: int, default=None  
  The number of instances in the test set.

## Forecasting Tags

- **is_univariate**: bool, default=True  
  Indicates whether the dataset is univariate. In the case of a forecasting dataset, this refers to the dimensionality of the `y` dataframe.
- **is_equally_spaced**: bool, default=True  
  Indicates whether all observations in the dataset are equally spaced.
- **has_nans**: bool, default=False  
  True if the dataset contains NaN values, False otherwise.
- **has_exogenous**: bool, default=False  
  True if the dataset contains exogenous variables, False otherwise.
- **n_instances**: int, default=None  
  The number of instances in the dataset. Should be equal to `y.shape[0]`.
- **n_instances_train**: int, default=None  
  The number of instances in the training set (None if the dataset does not have a train/test split). Should be equal to `y_train.shape[0]`.
- **n_instances_test**: int, default=None  
  The number of instances in the test set (None if the dataset does not have a train/test split). Should be equal to `y_test.shape[0]`.
- **n_timepoints**: int, default=None  
  The number of time points in the dataset, per series. If the dataset contains series of different lengths, this should be equal to the maximum length found in the dataset.
- **n_timepoints_train**: int, default=None  
  The number of time points in the training set, per series. If the dataset contains series of different lengths, this should be equal to the maximum length found in the training set.
- **n_timepoints_test**: int, default=None  
  The number of time points in the test set, per series. If the dataset contains series of different lengths, this should be equal to the maximum length found in the test set.
- **frequency**: str, default=None  
  The frequency of the time series in the dataset. Can be an integer if the frequency is not associated with a time unit.
- **n_dimensions**: int, default=1  
  The number of dimensions in the dataset (i.e., the number of columns in the `y` dataframe).
- **n_panels**: int, default=1  
  The number of panels in the dataset (i.e., the number of unique time series in the dataset).
- **n_hierarchy_levels**: int, default=0  
  The number of hierarchy levels in the dataset (i.e., the number of index levels in the `y` dataframe, excluding the time index).
  
  
 Any suggestions, @fkiraly ?

**Comment by fkiraly** [comment]
Looks brilliant, @felipeangelimvieira!

Might be worth opening a PR in the STEP repository, if it is not too much hassle.

Minor comments, but not blocking from my side and also not certain:

* if `X_train`, `y_train`, `X_test`, `y_test` end up generators, should we rename this, e.g., `X_trains`? Since this could cause confusion in the case of a single split. Although `X_trains` could also cause confusion, so I am undecided. Perhaps best leave as is?
* if `load` can return generators, the boilerplate should make sure that these are new generators each time, since generators are mutable.
* should there be a `scitype` or `scitypes` tag, or method? Although not certain what it should return, perhaps a dict `{"variable_name": "scitype_of_that_var"}`?

**Comment by fkiraly** [comment]
A couple more ideas:

* using the tag `name`, just as in the `datatypes` module. This is a shorthand and unique ID, similar as in the `huggingface` datasets namespace. Must be lower snake case.
* not necessary for first PR, introducing curated dataset collections. This is simply a list of datasets, and this could be versioned. For instance, Monash repository state 2023, state 2024, etc, as scientific reference points.

**Comment by felipeangelimvieira** [comment]
> if X_train, y_train, X_test, y_test end up generators, should we rename this, e.g., X_trains? Since this could cause confusion in the case of a single split. Although X_trains could also cause confusion, so I am undecided. Perhaps best leave as is?

I also wasn’t sure about the best approach in this scenario. "X_train" sounds more natural; however, it introduces inconsistent output types, which may require additional `if` and `else` statements in code leveraging this API—for benchmarking, for example. To maintain consistency across all scenarios, one solution could be to also return generators in the case of a single train-test split. This way, the user can always expect an iterator for train and test splits. This might be the best approach.

One drawback of keeping "X_train" instead of "X_trains" while ensuring consistent output is the requirement for extra steps like `X_train = next(X_train)` in tutorials:

```python
X_train, y_train = dataset.load("X_train", "y_train")
X_train = next(X_train)
y_train = next(y_train)
```

This conflicts with the common expectation of `X_train` being a DataFrame or an object that can be directly passed to an estimator. From this perspective, I’m starting to agree that "X_trains" could be a better solution for accessing CV folds’ data. In the case of a single train-test split, it could return generators of length 1. If there’s no train-test split and the user attempts to access "X_train," an exception could be raised, suggesting that the user retrieve data using "X_trains." Of course, the same logic applies to `y_train`.

> should there be a scitype or scitypes tag, or method? Although not certain what it should return, perhaps a dict {"variable_name": "scitype_of_that_var"}

Do you think this would be useful in addition to tags like `n_classes`, `n_panels`, and other metadata related to the content of the data? It could potentially help with documentation purposes.

> using the tag name, just as in the datatypes module. This is a shorthand and unique ID, similar as in the Hugging Face datasets namespace. Must be lower snake case. not necessary for first PR, introducing curated dataset collections. This is simply a list of datasets, and this could be versioned. For instance, Monash repository state 2023, state 2024, etc, as scientific reference points.

Agreed!

**Comment by felipeangelimvieira** [comment]
Franz, I've just created the step PR in sktime/enhancement-proposals#37.

I was thinking about "X_trains," "X_tests," etc., and also wondering if they might be too verbose. I believe users would rarely request one set without also needing the others, so having a single "cv" keyword might be better in that sense. Should we continue the discussion in the enhancement-proposals repo?

**Comment by fkiraly** [comment]
excellent, yes, will continue in the PR.
